File: ruby/GrabzIt/lib/grabzit/animationoptions.rb
module GrabzIt
	require File.join(File.dirname(__FILE__), 'baseoptions')
	
	# Represents all of the options available when creating an animated GIF
	# @version 3.0
	# @author GrabzIt
	class AnimationOptions < BaseOptions
		def initialize()
			@width = 0
			@height = 0
			@start = 0
			@duration = 0
			@speed = 0
			@framesPerSecond = 0
			@repeat = 0
			@reverse = false
			@customWaterMarkId = nil
			@quality = -1
		end
		
		# @return [Integer] the width of the resulting animated GIF in pixels
		def width
			@width
		end
		
		# Set the width of the resulting animated GIF in pixels
		#
		# @param value [Integer] the width
		# @return [void]		
		def width(value)
			@width = value
		end	

		# @return [Integer] the height of the resulting animated GIF in pixels
		def height
			@height
		end
		
		# Set the height of the resulting animated GIF in pixels
		#
		# @param value [Integer] the height
		# @return [void]		
		def height(value)
			@height = value
		end		

		# @return [Integer] the starting position of the video that should be converted into an animated GIF
		def start
			@start
		end
		
		# Set the starting position of the video that should be converted into an animated GIF
		#
		# @param value [Integer] the second to start at
		# @return [void]		
		def start(value)
			@start = value
		end		
		
		# @return [Integer] the length in seconds of the video that should be converted into a animated GIF
		def duration
			@duration
		end
		
		# Set the length in seconds of the video that should be converted into a animated GIF
		#
		# @param value [Integer] the number of seconds
		# @return [void]		
		def duration(value)
			@duration = value
		end				
		
		# @return [Float] the speed of the animated GIF
		def speed
			@speed
		end
		
		# Set the speed of the animated GIF from 0.2 to 10 times the original speed
		#
		# @param value [Float] the speed
		# @return [void]		
		def speed(value)
			@speed = value
		end					
		
		# @return [Float] the number of frames per second that should be captured from the video
		def framesPerSecond
			@framesPerSecond
		end
		
		# Set the number of frames per second that should be captured from the video. From a minimum of 0.2 to a maximum of 60
		#
		# @param value [Float] the number of frames per second
		# @return [void]		
		def framesPerSecond(value)
			@framesPerSecond = value
		end			
		
		# @return [Integer] the number of times to loop the animated GIF
		def repeat
			@repeat
		end
		
		# Set the number of times to loop the animated GIF. If 0 it will loop forever
		#
		# @param value [Integer] the number of times to loop
		# @return [void]		
		def repeat(value)
			@repeat = value
		end				
		
		# @return [Boolean] if the frames of the animated GIF should be reversed
		def reverse
			@reverse
		end
		
		# Set to true if the frames of the animated GIF should be reversed
		#
		# @param value [Boolean] reverse value
		# @return [void]		
		def reverse(value)
			@reverse = value
		end					
		
		# @return [String] the custom watermark id
		def customWaterMarkId
			@customWaterMarkId
		end
		
		# Set a custom watermark to add to the animated GIF
		#
		# @param value [String] custom watermark id
		# @return [void]		
		def customWaterMarkId(value)
			@customWaterMarkId = value
		end					
		
		# @return [Integer] the quality of the animated GIF
		def quality
			@quality
		end
		
		# Set the quality of the image where 0 is poor and 100 excellent. The default is -1 which uses the recommended quality
		#
		# @param value [Integer] the custom identifier
		# @return [void]		
		def quality(value)
			@quality = value
		end		
		
		# @!visibility private
		def _getSignatureString(applicationSecret, callBackURL, url = nil)
			urlParam = ''
			if (url != nil)
				urlParam = GrabzIt::Utility.nil_check(url)+"|"
			end		
			
			callBackURLParam = ''
			if (callBackURL != nil)
				callBackURLParam = GrabzIt::Utility.nil_check(callBackURL)
			end				
			
			return applicationSecret+"|"+ urlParam + callBackURLParam +
			"|"+GrabzIt::Utility.nil_int_check(@height)+"|"+GrabzIt::Utility.nil_int_check(@width)+"|"+GrabzIt::Utility.nil_check(@customId)+"|"+
			GrabzIt::Utility.nil_float_check(@framesPerSecond)+"|"+GrabzIt::Utility.nil_float_check(@speed)+"|"+GrabzIt::Utility.nil_int_check(@duration)+
			"|"+GrabzIt::Utility.nil_int_check(@repeat)+"|"+GrabzIt::Utility.b_to_str(@reverse)+"|"+GrabzIt::Utility.nil_int_check(@start)+
			"|"+GrabzIt::Utility.nil_check(@customWaterMarkId)+"|"+GrabzIt::Utility.nil_check(@country)+"|"+GrabzIt::Utility.nil_int_check(@quality)	  
		end		

		# @!visibility private
		def _getParameters(applicationKey, sig, callBackURL, dataName, dataValue)
			params = createParameters(applicationKey, sig, callBackURL, dataName, dataValue)	
			params['width'] = GrabzIt::Utility.nil_int_check(@width)
			params['height'] = GrabzIt::Utility.nil_int_check(@height)
			params['duration'] = GrabzIt::Utility.nil_int_check(@duration)
			params['speed'] = GrabzIt::Utility.nil_float_check(@speed)
			params['customwatermarkid'] = GrabzIt::Utility.nil_check(@customWaterMarkId)
			params['start'] = GrabzIt::Utility.nil_int_check(@start)
			params['fps'] = GrabzIt::Utility.nil_float_check(@framesPerSecond)
			params['repeat'] = GrabzIt::Utility.nil_int_check(@repeat)
			params['reverse'] = GrabzIt::Utility.b_to_str(@reverse)		
			params['quality'] = GrabzIt::Utility.nil_int_check(@quality)			
			
			return params
		end
	end
end

File: java/grabzit/src/it/grabz/grabzit/parameters/AnimationOptions.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.grabz.grabzit.parameters;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;

/**
 * Contains all of the available options when creating an animated GIF conversion.
 * 
 * @version 3.0
 * @author GrabzIt
 */
public class AnimationOptions extends BaseOptions{
    private int width;
    private int height;
    private int start;
    private int duration;
    private float speed;
    private float framesPerSecond;
    private int repeat;
    private boolean reverse;
    private String customWaterMarkId;
    private int quality; 

    public AnimationOptions()
    {
        this.width = 0;
        this.height = 0;
        this.start = 0;
        this.duration = 1;
        this.speed = 0;
        this.framesPerSecond = 0;
        this.repeat = 0;
        this.reverse = false;
        this.customWaterMarkId = "";
        this.quality = -1;
    }
    
    /**
     * @return the width of the resulting animated GIF in pixels.
     */
    public int getWidth() {
        return width;
    }

    /**
     * @param width the width of the resulting animated GIF in pixels.
     */
    public void setWidth(int width) {
        this.width = width;
    }

    /**
     * @return the height of the resulting animated GIF in pixels.
     */
    public int getHeight() {
        return height;
    }

    /**
     * @param height the height of the resulting animated GIF in pixels.
     */
    public void setHeight(int height) {
        this.height = height;
    }

    /**
     * @return the starting position of the video that should be converted into an animated GIF.
     */
    public int getStart() {
        return start;
    }

    /**
     * @param start the starting position of the video that should be converted into an animated GIF.
     */
    public void setStart(int start) {
        this.start = start;
    }

    /**
     * @return the length in seconds of the video that should be converted into a animated GIF.
     */
    public int getDuration() {
        return duration;
    }

    /**
     * @param duration the length in seconds of the video that should be converted into a animated GIF.
     */
    public void setDuration(int duration) {
        this.duration = duration;
    }

    /**
     * @return the the speed of the animated GIF.
     */
    public float getSpeed() {
        return speed;
    }

    /**
     * @param speed the speed of the animated GIF from 0.2 to 10 times the original speed.
     */
    public void setSpeed(float speed) {
        this.speed = speed;
    }

    /**
     * @return the number of frames per second that should be captured from the video.
     */
    public float getFramesPerSecond() {
        return framesPerSecond;
    }

    /**
     * @param framesPerSecond the number of frames per second that should be captured from the video. From a minimum of 0.2 to a maximum of 60.
     */
    public void setFramesPerSecond(float framesPerSecond) {
        this.framesPerSecond = framesPerSecond;
    }

    /**
     * @return the number of times to loop the animated GIF.
     */
    public int getRepeat() {
        return repeat;
    }

    /**
     * @param repeat the number of times to loop the animated GIF. If 0 it will loop forever.
     */
    public void setRepeat(int repeat) {
        this.repeat = repeat;
    }

    /**
     * @return if the frames of the animated GIF should be reversed.
     */
    public boolean isReverse() {
        return reverse;
    }

    /**
     * @param reverse set to true if the frames of the animated GIF should be reversed.
     */
    public void setReverse(boolean reverse) {
        this.reverse = reverse;
    }

    /**
     * @return the custom watermark id.
     */
    public String getCustomWaterMarkId() {
        return customWaterMarkId;
    }

    /**
     * @param customWaterMarkId a custom watermark to add to the animated GIF.
     */
    public void setCustomWaterMarkId(String customWaterMarkId) {
        this.customWaterMarkId = customWaterMarkId;
    }

    /**
     * @return the quality of the animated GIF.
     */
    public int getQuality() {
        return quality;
    }

    /**
     * @param quality the quality of the image where 0 is poor and 100 excellent. The default is -1 which uses the recommended quality.
     */
    public void setQuality(int quality) {
        this.quality = quality;
    }

    public String _getSignatureString(String applicationSecret, String callBackURL)
    {
        return _getSignatureString(applicationSecret, callBackURL, null);
    }    
    
    @Override    
    public String _getSignatureString(String applicationSecret, String callBackURL, String url)
    {
        String urlParam = "";
        if (url != null && !"".equals(url))
        {
            urlParam = ParameterUtility.nullCheck(url)+"|";
        }		

        String callBackURLParam = "";
        if (callBackURL != null && !"".equals(callBackURL))
        {
            callBackURLParam = ParameterUtility.nullCheck(callBackURL);
        }				

        return ParameterUtility.nullCheck(applicationSecret) + "|" + urlParam + callBackURLParam
        + "|" + height + "|" + width + "|" + getCustomId() + "|" + ParameterUtility.toString(framesPerSecond) 
        + "|" + ParameterUtility.toString(speed) + "|" + duration + "|" + repeat + "|" + ParameterUtility.toInt(reverse)
        + "|" + start + "|" + customWaterMarkId + "|" + getCountry().getValue() + "|" + quality + "|" + getExportURL()
        + "|" + getEncryptionKey() + "|" + getProxy();
    }    

    @Override
    public HashMap<String, String> _getParameters(String applicationKey, String sig, String callBackURL, String dataName, String dataValue) throws UnsupportedEncodingException
    {
        HashMap<String, String> params = createParameters(applicationKey, sig, callBackURL, dataName, dataValue);		
        params.put("width", String.valueOf(width));
        params.put("height", String.valueOf(height));
        params.put("duration", String.valueOf(duration));
        params.put("speed", ParameterUtility.toString(speed));
        params.put("customwatermarkid", ParameterUtility.nullCheck(customWaterMarkId));
        params.put("start", String.valueOf(start));
        params.put("fps", ParameterUtility.toString(framesPerSecond));
        params.put("repeat", String.valueOf(repeat));
        params.put("reverse", String.valueOf(ParameterUtility.toInt(reverse)));		
        params.put("quality", String.valueOf(quality));			

        return params;
    }    
}
